### PR TITLE
Migrate to Nunit4

### PR DESCRIPTION
**Pull request by jnyrup**
The first commit migrates from from NUnit3 to NUnit4.
* [Towards NUnit Version 4](https://docs.nunit.org/articles/nunit/Towards-NUnit4.html) explains what the new syntax offers.
* [Migration Guidance](https://docs.nunit.org/articles/nunit/release-notes/Nunit4.0-MigrationGuide.html) mentions breaking changes from NUnit3.

All the changes to the new `Assert.That` syntax (Called the constraint model) where done automatically by the added  [NUnit.Analyzers](https://github.com/nunit/nunit.analyzers).
Their list of [analyzers](https://github.com/nunit/nunit.analyzers/blob/master/documentation/index.md).

In the second commit I went through the warnings from [NUnit1028](https://github.com/nunit/nunit.analyzers/blob/master/documentation/NUnit1028.md) which warns about public methods in test classes that are _not_ marked with `[Test]`.
I've blindly added `[Test]` to all those that had asserts in them.

### CODE REVIEW

**Comment by kblok** [comment]
> All the changes to the new Assert.That syntax (Called the constraint model) where done automatically by the added [NUnit.Analyzers](https://github.com/nunit/nunit.analyzers).

I'm glad that you didn't have to do that manually!

**Comment by kblok** [comment]
I would try to land this soon. And we could work in some improvements later. 
I don't want to conflict this PR on every change.

**Comment by jnyrup** [comment]
@kblok the new NUnit analyzers pointed out methods that contained assertions, but weren't annotated with `[Test]`.
I've added those annotations in 9e188771ad120a25bf738f060a2761c89c38ddb6 but I could use help figuring out how they should be annotated.

File: README.md [hunk 1]
```diff
@@ -50,7 +50,7 @@ See [this document](https://github.com/hardkoded/puppeteer-sharp/blob/master/CON
 ## Take screenshots
 
 <!-- snippet: ScreenshotAsync -->
-<a id='snippet-ScreenshotAsync'></a>
+<a id='snippet-screenshotasync'></a>
```
Comment (kblok) [inline README.md:53]: I need to fix this. I bet you are in MacOS? I'm getting the same changes but the bot on github actions changes these back.

File: lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextOverridePermissionsTests.cs [hunk 2]
```diff
@@ -92,30 +90,31 @@ public async Task ShouldIsolatePermissionsBetweenBrowserContexts()
...
-            Assert.AreEqual("prompt", await GetPermissionAsync(Page, "geolocation"));
-            Assert.AreEqual("granted", await GetPermissionAsync(otherPage, "geolocation"));
+            Assert.That(await GetPermissionAsync(Page, "geolocation"), Is.EqualTo("prompt"));
+            Assert.That(await GetPermissionAsync(otherPage, "geolocation"), Is.EqualTo("granted"));
 
             await otherContext.CloseAsync();
         }
 
+        [Test, Ignore("previously not marked as a test")]
         public async Task AllEnumsdAreValid()
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextOverridePermissionsTests.cs:109]: Doing my homework and checking these tests.

```suggestion
        [Test]
        public async Task AllEnumsAreValid()
```

File: lib/PuppeteerSharp.Tests/ElementHandleTests/ContentFrameTests.cs [hunk 3]
```diff
@@ -22,17 +22,18 @@ public async Task ShouldWork()
             await FrameUtils.AttachFrameAsync(Page, "frame1", TestConstants.EmptyPage);
             var elementHandle = await Page.QuerySelectorAsync("#frame1");
             var frame = await elementHandle.ContentFrameAsync();
-            Assert.AreEqual(Page.FirstChildFrame(), frame);
+            Assert.That(frame, Is.EqualTo(Page.FirstChildFrame()));
         }
 
+        [Test, Ignore("previously not marked as a test")]
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/ElementHandleTests/ContentFrameTests.cs:28]:
```suggestion
        [Test]
```

File: README.md (hunk 1 above)
Comment (jnyrup) [reply README.md:53]: I'm on Windows and would really like if these could be fixed, I'm reverting them over and over

File: lib/PuppeteerSharp.Tests/FixturesTests/FixturesTests.cs [hunk 4]
```diff
@@ -28,9 +28,10 @@ public void ShouldDumpBrowserProcessStderr()
             process.Start();
             process.BeginErrorReadLine();
             process.WaitForExit();
-            Assert.True(success);
+            Assert.That(success, Is.True);
         }
 
+        [Test, Ignore("previously not marked as a test")]
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/FixturesTests/FixturesTests.cs:34]: Let's leave this as ignored.

File: lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextOverridePermissionsTests.cs (hunk 2 above)
Comment (jnyrup) [reply lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextOverridePermissionsTests.cs:109]: This test consistently fails for me in Firefox:

```
 AllEnumsdAreValid
   Source: BrowserContextOverridePermissionsTests.cs line 109
   Duration: 3,4 sec

  Message: 
PuppeteerSharp.MessageException : Protocol error (Browser.grantPermissions): Browser.grantPermissions RemoteAgentError@chrome://remote/content/cdp/Error.sys.mjs:18:5
UnknownMethodError@chrome://remote/content/cdp/Error.sys.mjs:101:7
execute@chrome://remote/content/cdp/domains/DomainCache.sys.mjs:92:13
execute@chrome://remote/content/cdp/sessions/Session.sys.mjs:59:25
onPacket@chrome://remote/content/cdp/CDPConnection.sys.mjs:257:36
onMessage@chrome://remote/content/server/WebSocketTransport.sys.mjs:127:18
handleEvent@chrome://remote/content/server/WebSocketTransport.sys.mjs:109:14


  Stack Trace: 
TaskHelper.WithTimeout[T](Task`1 task, TimeSpan timeout, Func`2 exceptionFactory) line 188
Connection.SendAsync(String method, Object args, Boolean waitForCallback, CommandOptions options) line 138
BrowserContextOverridePermissionsTests.AllEnumsdAreValid() line 112
GenericAdapter`1.BlockUntilCompleted()
NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaiter)
AsyncToSyncAdapter.Await[TResult](Func`1 invoke)
AsyncToSyncAdapter.Await(Func`1 invoke)
TestMethodCommand.RunTestMethod(TestExecutionContext context)
TestMethodCommand.Execute(TestExecutionContext context)
<>c__DisplayClass1_0.<Execute>b__0()
DelegatingTestCommand.RunTestMethodInThreadAbortSafeZone(TestExecutionContext context, Action action)

```

Comment (kblok) [reply lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextOverridePermissionsTests.cs:109]: Ah. I thought it was just a local check. Let's leave this as ignored. I'll write a rule to ignore it in Firefox in a follow-up PR.

File: lib/PuppeteerSharp.Tests/EvaluationTests/PageEvaluateOnNewDocumentTests.cs [hunk 5]
```diff
@@ -32,21 +32,22 @@ public async Task ShouldWorkWithCSP()
...
             // Make sure CSP works.
             await Page.AddScriptTagAsync(new AddTagOptions
             {
                 Content = "window.e = 10;"
             }).ContinueWith(_ => Task.CompletedTask);
-            Assert.Null(await Page.EvaluateExpressionAsync("window.e"));
+            Assert.That(await Page.EvaluateExpressionAsync("window.e"), Is.Null);
         }
 
+        [Test, Ignore("previously not marked as a test")]
```
[the remark this answers is not in the document]
Comment (jnyrup) [reply lib/PuppeteerSharp.Tests/EvaluationTests/PageEvaluateOnNewDocumentTests.cs:45]:
```
 ShouldWorkWithExpressions
   Source: PageEvaluateOnNewDocumentTests.cs line 46
   Duration: 2,2 sec

  Message: 
  Assert.That(await Page.EvaluateExpressionAsync<int>("window.result"), Is.EqualTo(123))
  Expected: 123
  But was:  0


  Stack Trace: 
PageEvaluateOnNewDocumentTests.ShouldWorkWithExpressions() line 50
GenericAdapter`1.BlockUntilCompleted()
NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaiter)
AsyncToSyncAdapter.Await[TResult](Func`1 invoke)
AsyncToSyncAdapter.Await(Func`1 invoke)
TestMethodCommand.RunTestMethod(TestExecutionContext context)
TestMethodCommand.Execute(TestExecutionContext context)
<>c__DisplayClass1_0.<Execute>b__0()
DelegatingTestCommand.RunTestMethodInThreadAbortSafeZone(TestExecutionContext context, Action action)
1)    at PuppeteerSharp.Tests.EvaluationTests.PageEvaluateOnNewDocumentTests.ShouldWorkWithExpressions() in C:\dev\puppeteer-sharp\lib\PuppeteerSharp.Tests\EvaluationTests\PageEvaluateOnNewDocumentTests.cs:line 50

```

[the remark this answers is not in the document]
Comment (kblok) [reply lib/PuppeteerSharp.Tests/EvaluationTests/PageEvaluateOnNewDocumentTests.cs:45]: Also firefox?
It passed on chrome on my side.

[the remark this answers is not in the document]
Comment (kblok) [reply lib/PuppeteerSharp.Tests/EvaluationTests/PageEvaluateOnNewDocumentTests.cs:45]: Ok.... Let's ignore it :( Follow up PR :(

**Comment by kblok** [comment]
@jnyrup do you think we can land this?

**Comment by jnyrup** [comment]
I don't have any pending changes. 

There's a failing tests with an NRE, I haven't looked into the cause of that, but I _guess_ it's unrelated to this PR.

**Pull request merged**